data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java

```diff
@@ -140,28 +141,36 @@ public static List<DataFile> listPartition(
             index -> {
               Metrics metrics = getAvroMetrics(fileStatus.get(index).getPath(), conf);
               datafiles[index] =
-                  buildDataFile(fileStatus.get(index), partitionValues, spec, metrics, "avro");
+                  metrics.recordCount() > 0
```

Comment:
Instead of adding null value when recordCount() gives zero, can't you skip adding anything in that case? Then you won't need to remove the null values before returning.